api/tests/iq_views_tests.py

```diff
@@ -0,0 +1,186 @@
+
+
+API_ROOT = "http://127.0.0.1:8000"
+INBOUND_SMS_PATH = f"{API_ROOT}/api/v1/inbound_sms_iq/"
+
+
+@pytest.fixture()
+def phone_user(db):
+    yield make_phone_test_user()
+
+
+@pytest.fixture(autouse=True)
+def mocked_twilio_client():
+    """
+    Mock PhonesConfig with a mock twilio client
+    """
+    with patch(
+        "phones.apps.PhonesConfig.twilio_client", spec_set=Client
+    ) as mock_twilio_client:
+        mock_fetch = Mock(
+            return_value=Mock(
+                country_code="US", phone_number="+12223334444", carrier="verizon"
+            )
+        )
+        mock_twilio_client.lookups.v1.phone_numbers = Mock(
+            return_value=Mock(fetch=mock_fetch)
+        )
+        yield mock_twilio_client
+
+
+@pytest.mark.django_db(transaction=True)
+def test_iq_endpoint_missing_verificationtoken_header():
+    client = RequestsClient()
+    response = client.post(INBOUND_SMS_PATH)
+    assert response.status_code == 401
+    response_body = response.json()
+    assert "missing Verificationtoken header" in response_body["detail"]
+
+
+@pytest.mark.django_db(transaction=True)
+def test_iq_endpoint_missing_messageid_header():
+    client = RequestsClient()
+    client.headers.update({"Verificationtoken": "valid"})
+    response = client.post(INBOUND_SMS_PATH)
+
+    assert response.status_code == 401
+    response_body = response.json()
+    assert "missing MessageId header" in response_body["detail"]
+
+
+@pytest.mark.django_db(transaction=True)
+def test_iq_endpoint_invalid_hash():
+    message_id = "9a09df23-01f3-4e0f-adbc-2a783878a574"
+    client = RequestsClient()
+    client.headers.update({"Verificationtoken": "invalid value"})
+    client.headers.update({"MessageId": message_id})
+    response = client.post(INBOUND_SMS_PATH)
+
+    assert response.status_code == 401
+    response_body = response.json()
+    assert "verficiationToken != computed sha256" in response_body["detail"]
+
+
+@pytest.mark.django_db(transaction=True)
+def test_iq_endpoint_valid_hash_no_auth_failed_status():
+    message_id = "9a09df23-01f3-4e0f-adbc-2a783878a574"
+    token = compute_iq_mac(message_id)
+    client = RequestsClient()
+    client.headers.update({"Verificationtoken": token})
+    client.headers.update({"MessageId": message_id})
+    response = client.post(INBOUND_SMS_PATH)
+
+    assert response.status_code == 400
+
+
+def _prepare_valid_iq_request_client() -> RequestsClient:
+    message_id = "9a09df23-01f3-4e0f-adbc-2a783878a574"
+    token = compute_iq_mac(message_id)
+    client = RequestsClient()
+    client.headers.update({"Verificationtoken": token})
+    client.headers.update({"MessageId": message_id})
+    return client
+
+
+@pytest.mark.django_db(transaction=True)
+def test_iq_endpoint_missing_required_params():
+    client = _prepare_valid_iq_request_client()
+
+    resp = client.post(INBOUND_SMS_PATH, {})
+
+    assert resp.status_code == 400
+    resp_body = resp.json()
+
+    # FIXME: why is this a list?
```

Comment:
Future work - I've noticed the same thing in other code. If we move parameter validation to a serializer, we'll get JSON objects like the model-backed endpoints. You can fake it by passing a `dict` to `ValidationError`, like:

```python
raise exceptions.ValidationError({"details": "Request missing from, to, or text.})
```
but converting to a serializer has benefits like documentation and reusing code.

Reply:
Filed a follow-up issue.